### PR TITLE
librbd: reduce mem copies to user-buffer during read

### PR DESCRIPTION
**Pull request by yuyuyu101**
rebase from #3399

### CODE REVIEW

File: src/common/buffer.cc [hunk 1]
```diff
@@ -1385,7 +1385,7 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
 	break;  // done
     }
   }
-  
+
```
Comment (tchaikov) [inline src/common/buffer.cc]: @yuyuyu101 where is `is_provided_buffer()` defined?

File: src/test/bufferlist.cc [hunk 2]
```diff
@@ -2698,6 +2698,15 @@ TEST(BufferList, InvalidateCrc) {
   EXPECT_NE(crc, bl.crc32c(0));
 }
 
+TEST(BufferList, TestIsProvidedBuffer) {
+  char buff[100];
+  bufferlist bl;
+  bl.push_back(buffer::create_static(100, buff));
+  ASSERT_EQ(bl.is_provided_buffer(buff), 1);
```
Comment (tchaikov) [inline src/test/bufferlist.cc]: might want use `ASSERT_TRUE()`

File: src/test/bufferlist.cc [hunk 3]
```diff
@@ -2698,6 +2698,15 @@ TEST(BufferList, InvalidateCrc) {
...
+  bl.append_zero(100);
+  ASSERT_EQ(bl.is_provided_buffer(buff), 0);
```
Comment (tchaikov) [inline src/test/bufferlist.cc]: and `ASSERT_FALSE()` here.

**Comment by yuyuyu101** [comment]
done

File: src/common/buffer.cc [hunk 4]
```diff
@@ -1315,6 +1315,12 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
     return true;
   }
 
+  bool buffer::list::is_provided_buffer(const char *dst) const
+  {
+    assert(!_buffers.empty());
```
Comment (dillaman) [inline src/common/buffer.cc]: Any particular reason for the assert?  Can't you just return false if the bl is empty?

**Comment by yuyuyu101** [comment]
@dillaman done

**Comment by jdurgin** [comment]
Ran into an issue with this while testing a python change on top - running this test ```PYTHONPATH=/home/jdurgin/ceph-master/src/test/pybind:/home/jdurgin/ceph-master/src/build/lib.linux-x86_64-2.7 LD_LIBRARY_PATH=/home/jdurgin/ceph-master/src/.libs nosetests -v test_rados:TestObject.test_read``` results in hitting an assert in c_str() via is_provided_buffer():

```
#0  0x00000031a7635c59 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00000031a7637368 in __GI_abort () at abort.c:89
#2  0x00007f64454a1015 in ceph::__ceph_assert_fail (assertion=assertion@entry=0x7f644571a480 "_raw", file=file@entry=0x7f644571a3da "common/buffer.cc", line=line@entry=810,
    func=func@entry=0x7f644571b200 <ceph::buffer::ptr::c_str() const::__PRETTY_FUNCTION__> "const char* ceph::buffer::ptr::c_str() const") at common/assert.cc:78
#3  0x00007f6445396bb3 in ceph::buffer::ptr::c_str (this=0x7fffd3193380) at common/buffer.cc:810
#4  0x00007f6445397693 in ceph::buffer::list::is_provided_buffer (this=this@entry=0x7fffd3193370, dst=dst@entry=0x27630e4 "TestObject.test_readh\255u\002") at common/buffer.cc:1320
#5  0x00007f64453afe33 in rados_read (io=0x282d680, o=o@entry=0x250c93c "foo", buf=buf@entry=0x27630e4 "TestObject.test_readh\255u\002", len=len@entry=100, off=off@entry=3) at librados/librados.cc:3098
#6  0x00007f644eabf455 in __pyx_pf_5rados_5Ioctx_46read (__pyx_v_offset=<optimized out>, __pyx_v_length=0x20852b0, __pyx_v_key=<optimized out>, __pyx_v_self=0x27640c0) at /home/jdurgin/ceph-master/src/build/rados.c:28660
#7  __pyx_pw_5rados_5Ioctx_47read (__pyx_v_self=0x27640c0, __pyx_args=<optimized out>, __pyx_kwds=<optimized out>) at /home/jdurgin/ceph-master/src/build/rados.c:28461

```

File: src/common/buffer.cc [hunk 5]
```diff
@@ -1315,6 +1315,11 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
     return true;
   }
 
+  bool buffer::list::is_provided_buffer(const char *dst) const
+  {
+    return (is_contiguous() && (_buffers.front().c_str() == dst));
```
Comment (dillaman) [inline src/common/buffer.cc]: Need ```if (_buffers.empty()) return false;```-like statement before attempting to dereference ```_buffers```

**Comment by yuyuyu101** [comment]
@dillaman @jdurgin done

**Pull request merged**